src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,129 @@
+(defpackage #:coalton-impl/codegen/constant-propagation
+  (:use
+   #:cl
+   #:coalton-impl/codegen/ast)
+  (:local-nicknames
+   (#:tc #:coalton-impl/typechecker))
+  (:import-from
+   #:coalton-impl/codegen/traverse
+   #:action
+   #:count-applications
+   #:count-nodes
+   #:make-traverse-let-action-skipping-cons-bindings
+   #:*traverse*
+   #:traverse
+   #:traverse-with-binding-list)
+  (:export
+   #:propagate-constants))
+
+(in-package #:coalton-impl/codegen/constant-propagation)
+
+(defun constant-var-value (var constant-bindings &key no-error)
+  (declare (type symbol var)
+           (values (or node null) &optional))
+  (let ((may-be-value (cdr (assoc var constant-bindings))))
+    (cond (may-be-value
+           ;; Whenever it exists, it is always a NODE (thus, non-NIL)
+           may-be-value)
+          (no-error
+           nil)
+          (t
+           (error "Expected VAR ~S to be a constant but is not" var)))))
+
+(defun propagate-constants (node env)
+  (declare (optimize debug))
+  (labels ((constant-node-p (node constant-bindings)
+             ;; FIXME: Do we need more nodes classified as constants?
+             (or (node-literal-p node)
+                 (and (node-variable-p node)
+                      (tc:lookup-instance-by-codegen-sym env (node-variable-value node) :no-error t))
+                 (and (node-variable-p node)
+                      (constant-var-value (node-variable-value node) constant-bindings :no-error t))))
+           (constant-node-value (node constant-bindings &key no-error)
+             ;; FIXME: Do we need more nodes classified as constants?
+             (cond ((node-literal-p node)
+                    node)
+                   ((and (node-variable-p node)
+                         (tc:lookup-instance-by-codegen-sym env (node-variable-value node) :no-error t))
+                    node)
+                   ((node-variable-p node)
+                    (constant-var-value (node-variable-value node) constant-bindings :no-error no-error))
+                   (no-error
+                    nil)
+                   (t
+                    (error "Expected the node to be a constant but is not ~%~S" node))))
+
+           (propagate-constants-node-variable (node constant-bindings)
+             (declare (type node-variable node))
+             (or (constant-node-value node constant-bindings :no-error t)
+                 node))
```

Comment:
All traversal actions will return the original node if the function returned `nil`, so this could also be written just as
```suggestion
             (constant-node-value node constant-bindings :no-error t)
```
(and at that point maybe having it be a separate function is unnecessary?)

Reply:
Implemented in https://github.com/coalton-lang/coalton/pull/1212/commits/37b690040c08e8f380871764e8458478d1404a58.